### PR TITLE
Made worker able to access sha

### PR DESCRIPTION
**Pull request by dolphingarlic**

### CODE REVIEW

File: src/worker.rs [hunk 1]
```diff
@@ -104,6 +108,18 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
...
+
+                                    let get_sha = Command::new("svn")
+                                        .arg("propget")
+                                        .arg("git-commit")
+                                        .arg("--revprop")
+                                        .arg("-r")
+                                        .arg(format!("{}", revision.unwrap()))
+                                        .arg(format!("{}/{}/trunk", ORGANIZATION_ROOT, query_name))
+                                        .output()
+                                        .expect("Failed to get sha of revision");
```
Comment (sushain97) [inline src/worker.rs]: Should never except this, then we'll panic and break the whole thread. Instead, match on it.

File: src/worker.rs [hunk 2]
```diff
@@ -104,6 +108,18 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
                                             err,
                                         )
                                     })?);
+
+                                    let get_sha = Command::new("svn")
```
Comment (sushain97) [inline src/worker.rs]: This will result in the command being called for every single revision, not every unique one.

File: src/worker.rs [hunk 3]
```diff
@@ -104,6 +108,18 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
...
+                                        .arg("propget")
+                                        .arg("git-commit")
+                                        .arg("--revprop")
+                                        .arg("-r")
+                                        .arg(format!("{}", revision.unwrap()))
```
Comment (sushain97) [inline src/worker.rs]: .to_string() suffices.

File: src/worker.rs (hunk 2 above)
Comment (dolphingarlic) [reply src/worker.rs]: Should I use a map that stores the unique revisions instead?

Comment (sushain97) [reply src/worker.rs]: What would the values be? Use a set.

Comment (dolphingarlic) [reply src/worker.rs]: So do I just not get the sha of a revision that already has a sha that was gotten?

File: src/worker.rs [hunk 4]
```diff
@@ -45,6 +46,40 @@ pub struct Worker {
     logger: Logger,
 }
 
+fn get_sha(
```
Comment (sushain97) [inline src/worker.rs]:
```suggestion
fn get_git_sha(
```

File: src/worker.rs [hunk 5]
```diff
@@ -45,6 +46,40 @@ pub struct Worker {
...
+    revision: Option<i32>,
```
Comment (sushain97) [inline src/worker.rs]:
```suggestion
    revision: i32,
```

File: src/worker.rs [hunk 6]
```diff
@@ -45,6 +46,40 @@ pub struct Worker {
...
+    query_name: &str,
```
Comment (sushain97) [inline src/worker.rs]:
```suggestion
    name: &str,
```

File: src/worker.rs [hunk 7]
```diff
@@ -45,6 +46,40 @@ pub struct Worker {
...
+    used_revisions: &mut HashMap<Option<i32>, Option<String>>
```
Comment (sushain97) [inline src/worker.rs]:
```suggestion
    revision_mapping: &mut HashMap<i32, String>
```

File: src/worker.rs [hunk 8]
```diff
@@ -45,6 +46,40 @@ pub struct Worker {
...
+                .arg("-r")
+                .arg(revision.unwrap().to_string())
+                .arg(format!("{}/{}/trunk", ORGANIZATION_ROOT, query_name))
+                .output();
+
+            match get_sha {
+                Ok(Output { status, ref stdout, .. }) if status.success() => {
+                    entry.insert(Some(format!("{}", String::from_utf8_lossy(stdout))));
+                },
+                Ok(Output { stderr: _, .. }) => {
```
Comment (sushain97) [inline src/worker.rs]: Instead, let's just return `None` and log an error.

File: src/worker.rs [hunk 9]
```diff
@@ -45,6 +46,40 @@ pub struct Worker {
...
+                .arg("-r")
+                .arg(revision.to_string())
+                .arg(format!("{}/{}/trunk", ORGANIZATION_ROOT, name))
+                .output();
+            
+            match get_sha {
+                Ok(Output { status, ref stdout, .. }) if status.success() => {
+                    entry.insert(Some(format!("{}", String::from_utf8_lossy(stdout))));
+                },
+                Ok(Output { stderr, .. }) => {
```
Comment (sushain97) [inline src/worker.rs]: I still think for both these issues we should be logging and then inserting `None`.

File: src/worker.rs [hunk 10]
```diff
@@ -45,6 +46,40 @@ pub struct Worker {
...
+                    entry.insert(Some(String::from("Unknown")));
+                },
+                Err(_) => {
+                    entry.insert(Some(String::from("Unknown")));
+                },
+            }
+        },
+        Occupied(_) => {},
```
Comment (sushain97) [inline src/worker.rs]: Instead, return the value you have here. and in the above branch, return a clone of the value you inserted. The `unwrap()` below is code smell. Avoid `unwrap()` and `expect()` as often as possible. We _do not_ want runtime panics.

File: src/worker.rs [hunk 11]
```diff
@@ -69,6 +104,8 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
         .arg(format!("{}/{}/trunk", ORGANIZATION_ROOT, name))
         .output();
 
+    let query_name = name;
```
Comment (sushain97) [inline src/worker.rs]: This isn't necessary. Cloning below might fix your issue.

File: src/worker.rs [hunk 12]
```diff
@@ -78,7 +115,10 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
             let mut files = Vec::new();
             let mut in_file_entry = false;
             let (mut in_name, mut in_author, mut in_date, mut in_size) = (false, false, false, false);
-            let (mut name, mut author, mut date, mut size, mut revision) = (None, None, None, None, None);
+            let (mut name, mut author, mut date, mut size, mut revision, mut sha) =
+                (None, None, None, None, None, None);
+
+            let mut revision_mapping: HashMap<i32, Option<String>> = HashMap::new();
```
Comment (sushain97) [inline src/worker.rs]: Move this to after the `let mut files`...?

File: src/worker.rs (hunk 11 above)
Comment (dolphingarlic) [reply src/worker.rs]: Doesn't name get turned into None later on though?

File: migrations/2018-12-04-064410_added_sha_to_table/down.sql [hunk 13]
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE entries_backup(
```
Comment (sushain97) [inline migrations/2018-12-04-064410_added_sha_to_table/down.sql]: rename to `add_sha_to_entries` to match the others

File: .vscode/launch.json [hunk 14]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (sushain97) [inline .vscode/launch.json]: Don't commit this file!

File: src/worker.rs [hunk 15]
```diff
@@ -45,6 +49,46 @@ pub struct Worker {
     logger: Logger,
 }
 
+fn get_git_sha(
+    revision: i32,
```
Comment (sushain97) [inline src/worker.rs]: Just for consistency, let's make the order of arguments here, `logger`, `revision_mapping`, `revision`, `svn_path`.

File: src/worker.rs [hunk 16]
```diff
@@ -45,6 +49,46 @@ pub struct Worker {
...
+
+            match get_sha {
+                Ok(Output { status, ref stdout, .. }) if status.success() => {
+                    let sha = Some(format!("{}", String::from_utf8_lossy(stdout)));
+                    entry.insert(sha.clone());
+                    sha
+                },
+                Ok(Output { stderr, .. }) => {
+                    let err = String::from_utf8_lossy(&stderr);
+                    error!(logger, "Cannot get sha of revision: {}", err);
```
Comment (sushain97) [inline src/worker.rs]: Here and below:

```suggestion
                    error!(logger, "Error getting SHA corresponding to revision: {:?}", err, "revision" => revision);
```

File: src/worker.rs [hunk 17]
```diff
@@ -45,6 +49,46 @@ pub struct Worker {
...
+                .arg("git-commit")
+                .arg("--revprop")
+                .arg("-r")
+                .arg(revision.to_string())
+                .arg(svn_path)
+                .output();
+
+            match get_sha {
+                Ok(Output { status, ref stdout, .. }) if status.success() => {
+                    let sha = Some(format!("{}", String::from_utf8_lossy(stdout)));
```
Comment (sushain97) [inline src/worker.rs]: I think you can use `clone()` here instead of `format!`

File: src/worker.rs [hunk 18]
```diff
@@ -69,16 +113,20 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
         .arg(format!("{}/{}/trunk", ORGANIZATION_ROOT, name))
         .output();
 
+    let svn_path = format!("{}/{}/trunk", ORGANIZATION_ROOT, name);
```
Comment (sushain97) [inline src/worker.rs]: Let's use `svn_path` above as well.

File: src/worker.rs [hunk 19]
```diff
@@ -104,6 +152,9 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
                                             err,
                                         )
                                     })?);
+
+                                    sha = get_git_sha(revision.unwrap(), &svn_path, &mut revision_mapping, &logger);
```
Comment (sushain97) [inline src/worker.rs]: No unwraps!

File: src/worker.rs [hunk 20]
```diff
@@ -97,13 +151,19 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
...
                             if let Ok(Attribute { value, key }) = attr {
                                 if decode_utf8(&key, &reader)? == "revision" {
-                                    revision = Some(decode_utf8(&value, &reader)?.parse::<i32>().map_err(|err| {
-                                        format!(
-                                            "Revision number parsing error at position {}: {:?}",
-                                            reader.buffer_position(),
-                                            err,
-                                        )
-                                    })?);
+                                    let unwrapped_revision =
```
Comment (sushain97) [inline src/worker.rs]: unwrapped is a bit overloaded. how about,
```suggestion
                                    let raw_revision =
```

or `revision_str`.

File: src/worker.rs [hunk 21]
```diff
@@ -97,13 +151,19 @@ fn list_files(logger: &Logger, name: &str, recursive: bool) -> Result<Vec<File>,
...
+                                        decode_utf8(&value, &reader)?.parse::<i32>().map_err(|err| {
+                                            format!(
+                                                "Revision number parsing error at position {}: {:?}",
+                                                reader.buffer_position(),
+                                                err,
+                                            )
+                                        })?;
+
+                                    sha = get_git_sha(&logger, &mut revision_mapping, unwrapped_revision, &svn_path);
+
```
Comment (sushain97) [inline src/worker.rs]:
```suggestion
```

**Pull request merged**